### ISSUE TITLE
feat request - cmake toolchain template

### ISSUE THREAD

**Issue by RodrigoDornelles** [cc65/cc65#1884]
It would be interesting as there are already default settings to be used, a build cmake toolchain for cc65, as well as a project of the same purpose already has <https://github.com/llvm-mos/llvm-mos-sdk>

makes it easier for larger projects to do an automated build.
 
* PS. if you approve I'll be available to try to implement this, I'm still learning both about cmake and cc65, but I think it would be a useful contribution.

**Comment by mrdudz** [comment]
Not sure what exactly the request is - but it sounds like this is out of scope and may belong into the wiki - just like that universal makefile someone created

**Comment by RodrigoDornelles** [comment]
basically this <https://github.com/llvm-mos/llvm-mos-sdk/blob/main/cmake/llvm-mos-toolchain.cmake> to easily compile cross-platform projects using gnu cmake.

to use like this

```cmake
cmake_minimum_required(VERSION 3.18)
set(CC65_MOS_PLATFORM c64)
set(CMAKE_TOOLCHAIN_FILE ./cc65/cmake/toolchain.cmake)
add_executable(foo foo.c)
```

it's not like a "unviersal makefile", but tools to build your own "CMakeList.txt" and generate a makefile, I don't see why it would be outside the scope of the project.

**Comment by rofl0r** [comment]
one can use any compiler with regular buildsystems by setting the CC variable, e.g. `make CC=cc65`. this should even work with cmake, even though it tends to be quite inflexible (result of being developed by people that didn't understand the [problem space](https://sabotage-linux.neocities.org/blog/5/)), but not with meson where everything needs to be hardcoded: https://github.com/mesonbuild/meson/issues/5406

**Comment by RodrigoDornelles** [comment]
of course you can use cc65 with cmake/makefile, but I'm suggesting provide a cmake toolchain, this will make it much easier for programmers to create automated projects with cc65 compiler.

**Comment by groessler** [comment]
https://twdev.blog/2021/08/cmake/
:-)

**Comment by groessler** [comment]
Why are the current Makefiles bad?

**Comment by RodrigoDornelles** [comment]
> https://twdev.blog/2021/08/cmake/ :-)

I particularly don't like cmake, but it's a common pattern in the C community, and mainly a common adoption between IDE's

Of course cmake is kind of boring, but useful, and the opinion of this post is something very personal.<br/>_(still influenced by meson)_

----------------------

> Why are the current Makefiles bad?

I'm not talking about changing the cc65 project from makefile to cmake, but adding builder helper to your own project.

#### folder structure ###

```ruby
my-awesome-project/
  CMakeLists.txt
  src/
   ... #source code project
  cc65/ # submodule
     Makefile #actual makefile
     cmake/ # new folder
        CC65_toolchain.cmake # my proposal
```

#### files ####

`CMakeLists.txt`
```
set(CC65_MOS_PLATFORM c64)
set(CMAKE_TOOLCHAIN_FILE ./cc65/cmake/CC65_toolchain.cmake)
add_executable(foo foo.c)
```

`CC65_toolchain.cmake`

 * <https://github.com/llvm-mos/llvm-mos-sdk/blob/main/cmake/llvm-mos-toolchain.cmake>
 * <https://github.com/jviskari/cc65_cmake/blob/master/toolchain/cc65-toolchain.cmake>

#### comands ####

```shell
$ git clone https://example.com/someone/my-awesome-project --recursive
$ cd my-awesome-project
$ cmake .
$ make
```

**Comment by polluks** [comment]
cc65 is a _clean_ ANSI C project.
If you need build automation right now, something goes wrong.
I don't want to go into CMake hell anyway.

**Comment by jroweboy** [comment]
@polluks Neither I nor anyone else are suggesting that we add cmake to the cc65 project, I think we all agree cc65 doesn't need cmake. To be clear, a cmake-toolchain file is for the USERS of the cc65 compiler should they want to use cmake. This is why I'm suggesting the toolchain can live as a community addition on the wiki.

**Comment by RodrigoDornelles** [comment]
here's a good example of a ziglog project, it's interesting that compilers provide toolchains to make it easier for homebrew creators to work cross-platform easily. if you allow it we can put it in the cc65 project without disturbing the source code.

 * https://github.com/z88dk/z88dk/blob/master/support/cmake/Toolchain-zcc.cmake

**Comment by mrdudz** [comment]
The biggest problem i see with it is: this stuff needs to be maintained (and ideally tested in the CI). I am not using cmake myself, unless getting paid for it :) The documentation really really sucks to begin with, and i couldnt answer any questions about it either.

**Comment by RodrigoDornelles** [comment]
It might be something I can maintain (including testing it in CI), am depend on CC65 for my own interpreted language and just getting started with support.

I don't like to use cmake, but sometimes it's necessary to cry a little :o

**Comment by sidneycadot** [comment]
May be a silly question, but shouldn't CMake support be added to the CMake project, rather than to the cc65 project?

**Comment by polluks** [comment]
@sidneycadot Possible. CMake could manage cc65, vbcc6502 and Oscar64 in one project.
